sslcontext-kickstart/src/main/java/nl/altindag/ssl/trustmanager/InflatableX509ExtendedTrustManager.java

```diff
@@ -39,20 +52,95 @@ public class InflatableX509ExtendedTrustManager extends HotSwappableX509Extended
         super(TrustManagerUtils.createDummyTrustManager());
 
         writeLock.lock();
 
         try {
-            trustStore = KeyStoreUtils.createKeyStore();
+            this.trustStorePath = trustStorePath;
+            this.trustStorePassword = trustStorePassword;
+            this.trustStoreType = trustStoreType;
+
+            this.certificateAndAuthTypeTrustPredicate = Optional.ofNullable(certificateAndAuthTypeTrustPredicate)
+                    .orElse((chain, authType) -> false);
+
+            if (trustStorePath != null && trustStorePassword != null && StringUtils.isNotBlank(trustStoreType)) {
+                if (Files.exists(trustStorePath)) {
+                    trustStore = KeyStoreUtils.loadKeyStore(trustStorePath, trustStorePassword, trustStoreType);
+                } else {
+                    trustStore = KeyStoreUtils.createKeyStore(trustStoreType, trustStorePassword);
+                }
+            } else {
+                trustStore = KeyStoreUtils.createKeyStore();
+            }
         } finally {
             writeLock.unlock();
         }
     }
 
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        checkTrusted(trustManager -> super.checkServerTrusted(chain, authType), chain, authType);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        checkTrusted(trustManager -> super.checkServerTrusted(chain, authType, socket), chain, authType);
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine) throws CertificateException {
+        checkTrusted(trustManager -> super.checkServerTrusted(chain, authType, sslEngine), chain, authType);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+        checkTrusted(trustManager -> super.checkClientTrusted(chain, authType), chain, authType);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket) throws CertificateException {
+        checkTrusted(trustManager -> super.checkClientTrusted(chain, authType, socket), chain, authType);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine) throws CertificateException {
+        checkTrusted(trustManager -> super.checkClientTrusted(chain, authType, sslEngine), chain, authType);
+    }
+
+    private void checkTrusted(TrustManagerConsumer trustManagerConsumer, X509Certificate[] chain, String authType) throws CertificateException {
+        readLock.lock();
+
+        try {
+            trustManagerConsumer.checkTrusted(this);
+        } catch (CertificateException e) {
+            boolean shouldBeTrusted = certificateAndAuthTypeTrustPredicate.test(chain, authType);
```

Comment:
As far as I know the readlock can be acquired by multiple threads as long as they don't change the data. The writer lock can be acquired by a single thread and others need to wait till it is finished. So coming back to your concern, I think the predicate can be called by multiple different threads, but that is something which we should prevent, right?

You want to provide the user a single pop-up with asking to trust the certificate at a time, right?

Reply:
In case you want to avoid the synchronize block, I made an attempt to use double checked locking: https://github.com/Hakky54/sslcontext-kickstart/pull/342/commits/5d1556d2d1a07e0ed534d3611cc37f4299f56a94

Also I confirm that trying to upgrade a read lock to a write lock will cause dead locks, so a read lock should always be released before trying to take a write lock.